codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -541,6 +537,10 @@ private boolean shouldIgnoreHeadersOrDataFrame(ChannelHandlerContext ctx, int st
                                 ("Stream created after GOAWAY sent. Last known stream by peer " +
                                  connection.remote().lastStreamKnownByPeer()));
                 }
+
+                // If we are ignoring frames for this stream it means that we no longer need to retain state for this
+                // stream and can close it, which should remove it from the active stream map.
+                stream.close();
```

Comment:
It doesn't seem like the responsibility of the decoder to manage the streams life-cycle Under what conditions would this happen? It seems that sending a RST should always be coupled with closing the stream and you've covered the GOAWAY case above. From my limited digging, it looks like the [`Http2ConnectionHandler`](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java#L778) is the only way we'll hit the `isResetSent()` case and it's already setup a listener to close the session. I presume it's only delayed to make draining work correctly with resets.

Reply:
@Scottmitch can you chime in here ?

Reply:
This was added as a fail-safe because prior to this PR we would still create streams initiated by the remote endpoint even if we sent an goaway. This resulted in "leaked" stream objects (stream objects would be created, inserted into the `streamMap` but never removed because there would be no lifetime event to trigger removal). With this PR `DefaultHttp2Connection` no longer permits this and I don't think we should hit this code path because:

1. as you said sending a rst should close the stream, which will remove from the `streamMap`.
2. we should no longer create streams initiated by the remote peer (exceeding the last stream ID) after a go away was sent.

so I can remove this code block if we assume this scenario is invalid for any implementation of `Http2Connection`. We have unit tests which do exercise this code path in `DefaultHttp2ConnectionDecoderTest` but we can modify the test/mocks to not

Reply:
I can't attest to other implementations of `Http2Connection`, but if that's a legit concern I don't mind if we leave it if the worst case scenario is it's dead code. I'll trust your judgment @Scottmitch.

Reply:
I'm not aware of any alternative implementation of `Http2Connection` and you have a good point this code change seems to be fixing an issue that stems from elsewhere. Now that `DefaultHttp2Connection` is fixed it should be sufficient to remove this with Netty's code, but I'll add a comment to clarify the assumptions.